File: classes/ranger.py
from typing import List, override, Optional

import sim.core_feats
from sim.events import DamageRollArgs
from util.util import get_magic_weapon, apply_asi_feats
from feats.epic_boons import IrresistibleOffense
from feats.fighting_style import TwoWeaponFighting, Archery
from feats import (
    ASI,
    CrossbowExpert,
    DualWielder,
    WeaponMasteries,
    LightWeaponBonusAttack,
)
from weapons import HandCrossbow, Shortsword, Scimitar, Rapier
from spells.ranger import HuntersMark
from spells.summons import SummonFey
from sim.spells import Spellcaster


import sim.feat
import sim.character
import sim.target
import sim.weapons


def maybe_cast_summon_fey(
    character: "sim.character.Character",
    summon_fey_threshold: int,
):
    slot = character.spells.highest_slot()
    if slot >= summon_fey_threshold and not character.spells.is_concentrating():
        spell = SummonFey(slot)
        character.spells.cast(spell)
        return True
    return False


def maybe_cast_hunters_mark(
    character: "sim.character.Character",
    target: "sim.target.Target",
):
    slot = character.spells.lowest_slot()
    if (
        slot > 0
        and not character.spells.is_concentrating()
        and character.use_bonus("HuntersMark")
    ):
        character.spells.cast(HuntersMark(slot), target)
        return True
    return False


class RangerLevel(sim.core_feats.ClassLevels):
    def __init__(self, level: int):
        super().__init__(name="Ranger", level=level, spellcaster=Spellcaster.HALF)


class BestialFury(sim.feat.Feat):
    def __init__(
        self,
        caster: Optional["sim.character.Character"] = None,
    ) -> None:
        self.caster = caster

    def apply(self, character: "sim.character.Character"):
        super().apply(character)
        if self.caster is None:
            self.caster = character

    def attack_result(self, args):
        if args.hits() and args.attack.target.has_tag("HuntersMark"):
            args.add_damage(
                source="HuntersMark",
                dice=[6],
            )


class FoeSlayer(sim.feat.Feat):
    def damage_roll(self, args: DamageRollArgs):
        if args.damage.source == "HuntersMark":
            for i in range(len(args.damage.dice)):
                args.damage.dice[i] = 8
            args.damage.reroll()


class PreciseHunter(sim.feat.Feat):
    def attack_roll(self, args):
        if args.attack.target.has_tag("HuntersMark"):
            args.adv = True


class DreadAmbusher(sim.feat.Feat):
    def __init__(self, level: int) -> None:
        self.die = 8 if level >= 11 else 6
        self.uses = 0

    def long_rest(self):
        self.uses = self.character.mod("wis")

    def begin_turn(self, target: "sim.target.Target"):
        self.used = False

    def attack_result(self, args):
        if args.hits() and not self.used and self.uses > 0:
            self.used = True
            self.uses -= 1
            args.add_damage(source="DreadAmbusher", dice=2 * [self.die])


class BeastChargeFeat(sim.feat.Feat):
    def __init__(self, character: "sim.character.Character"):
        self.character = character

    def attack_result(self, args):
        if args.hits() and isinstance(args.attack.weapon, BeastMaul):
            args.add_damage(source="Charge", dice=[6])
            if not args.attack.target.prone and not args.attack.target.save(
                self.character.dc("wis")
            ):
                args.attack.target.knock_prone()


def ranger_feats(
    level: int,
    masteries: List["sim.weapons.WeaponMastery"],
    fighting_style: "sim.feat.Feat",
    asis: Optional[List["sim.feat.Feat"]] = None,
) -> List["sim.feat.Feat"]:
    feats: List["sim.feat.Feat"] = []
    if level >= 1:
        feats.append(RangerLevel(level))
        feats.append(WeaponMasteries(masteries))
    # Level 2 (Deft Explorer) is irrelevant
    if level >= 2:
        feats.append(fighting_style)
    # Level 6 (Roving) is irrelevant
    # Level 9 (Expertise) is irrelevant
    # Level 10 (Tireless) is irrelevant
    # Level 13 (Relentless Hunter) is irrelevant
    # TODO: Level 14 (Nature's Veil)
    if level >= 17:
        feats.append(PreciseHunter())
    # Level 18 (Feral Senses) is irrelevant
    if level >= 20:
        feats.append(FoeSlayer())
    apply_asi_feats(level=level, feats=feats, asis=asis)
    return feats


def gloomstalker_ranger_feats(
    level: int, weapon: "sim.weapons.Weapon"
) -> List["sim.feat.Feat"]:
    feats: List["sim.feat.Feat"] = []
    if level >= 3:
        feats.append(DreadAmbusher(level))
    # Level 3 (Umbral Sight) is irrelevant
    # Level 7 (Iron Mind) is irrelevant
    # Level 11 (Stalker's Flurry) is handled elsewhere
    # Level 15 (Shadowy Dodge) is irrelevant
    return feats


class GloomstalkerAction(sim.feat.Feat):
    def __init__(
        self, attacks: List["sim.weapons.Weapon"], summon_fey_threshold: int
    ) -> None:
        self.attacks = attacks
        self.summon_fey_threshold = summon_fey_threshold

    def action(self, target: "sim.target.Target"):
        if maybe_cast_summon_fey(self.character, self.summon_fey_threshold):
            return
        maybe_cast_hunters_mark(self.character, target)
        for weapon in self.attacks:
            self.character.weapon_attack(target, weapon, tags=["main_action"])
        self.character.weapon_attack(target, self.attacks[0], tags=["light"])


class GloomstalkerRanger(sim.character.Character):
    def __init__(self, level: int):
        magic_weapon = get_magic_weapon(level)
        feats: List["sim.feat.Feat"] = []
        weapon: "sim.weapons.Weapon" = HandCrossbow(magic_bonus=magic_weapon)
        attacks: List["sim.weapons.Weapon"] = [weapon]
        if level >= 5:
            attacks = 2 * [weapon]
        feats.append(GloomstalkerAction(attacks=attacks, summon_fey_threshold=4))
        feats.extend(
            ranger_feats(
                level=level,
                masteries=["Vex", "Nick"],
                fighting_style=Archery(),
                asis=[
                    CrossbowExpert(weapon),
                    ASI(["dex"]),
                    ASI(["wis"]),
                    ASI(["wis"]),
                    IrresistibleOffense("dex"),
                ],
            )
        )
        feats.extend(gloomstalker_ranger_feats(level, weapon))
        super().init(
            level=level,
            stats=[10, 17, 10, 10, 16, 10],
            base_feats=feats,
            spell_mod="wis",
        )


class PrimalCompanion(sim.character.Character):
    def __init__(self, level: int, ranger: "sim.character.Character"):
        self.ranger = ranger
        self.num_attacks = 2 if level >= 11 else 1
        self.weapon = BeastMaul(self)
        feats: List["sim.feat.Feat"] = []
        feats.append(BeastChargeFeat(ranger))
        if level >= 11:
            feats.append(BestialFury(caster=ranger))
        if level >= 20:
            feats.append(FoeSlayer())
        super().init(
            level=level,
            stats=[10, 10, 10, 10, 10, 10],
            base_feats=feats,
        )

    def do_attack(self, target: "sim.target.Target"):
        for _ in range(self.num_attacks):
            self.weapon_attack(target, self.weapon)


class BeastMaul(sim.weapons.Weapon):
    def __init__(self, ranger: "sim.character.Character", **kwargs):
        super().__init__(name="Beast Maul", num_dice=1, die=8, **kwargs)
        self.ranger = ranger

    @override
    def to_hit(self, character):
        return self.ranger.spells.to_hit()

    @override
    def attack_result(self, args, character):
        if not args.hits():
            return
        num_dice = 2 if args.crit else 1
        args.add_damage(
            source=self.name,
            dice=num_dice * [8],
            damage=2 + self.ranger.mod("wis"),
        )


class BeastMasterAction(sim.feat.Feat):
    def __init__(
        self,
        level: int,
        beast: PrimalCompanion,
        magic_bonus: int,
    ) -> None:
        self.beast = beast
        self.shortsword = Shortsword(magic_bonus=magic_bonus)
        self.scimitar = Scimitar(magic_bonus=magic_bonus)
        if level >= 4:
            self.main_hand: "sim.weapons.Weapon" = Rapier(magic_bonus=magic_bonus)
        else:
            self.main_hand = Shortsword(magic_bonus=magic_bonus)

    def action(self, target: "sim.target.Target"):
        if maybe_cast_summon_fey(self.character, summon_fey_threshold=4):
            return
        maybe_cast_hunters_mark(self.character, target)
        if self.character.level >= 3:
            commanded_beast = False
            if self.character.use_bonus("beast"):
                self.beast.do_attack(target)
                commanded_beast = True
            num_attacks = 2 if self.character.level >= 5 else 1
            for _ in range(num_attacks):
                if not commanded_beast:
                    self.beast.do_attack(target)
                    commanded_beast = True
                else:
                    self.character.weapon_attack(target, self.main_hand)
            self.character.weapon_attack(target, self.shortsword, tags=["light"])
        else:
            self.character.weapon_attack(target, self.main_hand)
            # Use bonus action to use Vex weapon, if possible
            if self.character.use_bonus("light weapon"):
                self.character.weapon_attack(target, self.scimitar, tags=["light"])
            else:
                self.character.weapon_attack(target, self.shortsword, tags=["light"])


def beast_master_ranger_feats(level: int) -> List["sim.feat.Feat"]:
    feats: List["sim.feat.Feat"] = []
    # TODO: Move Primal Companion into here
    # Level 7 (Exceptional Training) is irrelevant
    # Level 11 (Bestial Fury) is handled elsewhere
    # TODO: Level 15 (Share Spells)
    return feats


class BeastMasterRanger(sim.character.Character):
    def __init__(self, level: int):
        magic_weapon = get_magic_weapon(level)
        feats: List["sim.feat.Feat"] = []
        beast = PrimalCompanion(level, ranger=self)
        shortsword = Shortsword(magic_bonus=magic_weapon)
        feats.append(
            BeastMasterAction(
                level=level,
                beast=beast,
                magic_bonus=magic_weapon,
            )
        )
        feats.extend(
            ranger_feats(
                level=level,
                masteries=["Vex", "Nick"],
                fighting_style=TwoWeaponFighting(),
                asis=[
                    DualWielder("dex", shortsword),
                    ASI(["dex"]),
                    ASI(["wis"]),
                    ASI(["wis"]),
                    IrresistibleOffense("dex"),
                ],
            )
        )
        feats.extend(beast_master_ranger_feats(level))
        super().init(
            level=level,
            stats=[10, 17, 10, 10, 16, 10],
            base_feats=feats,
            spell_mod="wis",
        )
        self.add_minion(beast)

File: src/classes/Ranger.ts
import { WeaponMastery } from "../sim/types"
import { Feature } from "../sim/Feature"
import { AddClassLevel } from "../sim/coreFeats/ClassLevel"
import { WeaponMasteries } from "../feats/shared/WeaponMasteries"
import { defaultMagicBonus, unreachable } from "../util/helpers"
import { Character } from "../sim/Character"
import { ExtraAttack } from "../feats/shared/ExtraAttack"
import { DamageRollEvent } from "../sim/events/DamageRollEvent"
import { AttackRollEvent } from "../sim/events/AttackRollEvent"
import { Archery } from "../feats/fightingStyle/Archery"
import { AbilityScoreImprovement } from "../feats/general/AbilityScoreImprovement"
import { IrresistibleOffense } from "../feats/epic/IrresistibleOffense"
import { DefaultAttackActionOperation } from "../operations/DefaultAttackActionOperation"
import { Longbow } from "../weapons/martial/ranged/Longbow"
import { CastSpellOperation } from "../operations/CastSpellOperation"
import { SummonFey } from "../spells/SummonFey"
import { HuntersMark } from "../spells/HuntersMark"
import { Spellcaster } from "../sim/spells/shared"
import { SpellcastingFeat } from "../sim/spells/SpellcastingFeat"
import { BeginTurnEvent } from "../sim/events/BeginTurnEvent"
import { AttackResultEvent } from "../sim/events/AttackResultEvent"
import { ShortRestEvent } from "../sim/events/ShortRestEvent"
import { FeatureGroup } from "../sim/helpers/FeatureGroup"

class PreciseHunter extends Feature {
    attackRoll(event: AttackRollEvent): void {
        if (this.character.hasEffect("HuntersMark")) {
            event.adv = true
        }
    }
}

class FoeSlayer extends Feature {
    damageRoll(event: DamageRollEvent): void {
        if (event.damage.source === "HuntersMark") {
            event.damage.replaceDice([10])
        }
    }
}

class ColossusSlayer extends Feature {
    used: boolean = false
    doneDamage: boolean = false

    shortRest(event: ShortRestEvent): void {
        this.used = false
        this.doneDamage = false
    }

    beginTurn(event: BeginTurnEvent): void {
        this.used = false
    }

    attackResult(event: AttackResultEvent): void {
        const weapon = event.attack.attack.weapon()
        if (event.hit && !this.used && weapon && this.doneDamage) {
            this.used = true
            event.addDamage({
                source: "ColossusSlayer",
                type: weapon.damageType,
                dice: [8],
            })
        }
    }

    damageRoll(event: DamageRollEvent): void {
        this.doneDamage = true
    }
}

type RangerSubclass = "Hunter"

export class Ranger {
    static features(args: {
        level: number
        asis: Feature[]
        masteries: WeaponMastery[]
        fightingStyle: Feature
        subclass: RangerSubclass
    }): Feature[] {
        const { level, asis, masteries, fightingStyle, subclass } = args
        const features: Feature[] = []
        if (level >= 1) {
            features.push(Ranger.level1(level, masteries))
        }
        // Level 2 (Deft Explorer) is irrelevant
        if (level >= 2) {
            features.push(Ranger.level2(fightingStyle))
        }
        if (level >= 3) {
            features.push(Ranger.level3(subclass))
        }
        if (level >= 4) {
            features.push(Ranger.level4(asis[0]))
        }
        if (level >= 5) {
            features.push(Ranger.level5())
        }
        // Level 6 (Roving) is irrelevant
        if (level >= 8) {
            features.push(Ranger.level8(asis[1]))
        }
        // Level 9 (Expertise) is irrelevant
        // Level 10 (Tireless) is irrelevant
        if (level >= 12) {
            features.push(Ranger.level12(asis[2]))
        }
        // Level 13 (Relentless Hunter) is irrelevant
        // TODO: Level 14 (Nature's Veil)
        if (level >= 16) {
            features.push(Ranger.level16(asis[3]))
        }
        if (level >= 17) {
            features.push(Ranger.level17())
        }
        // Level 18 (Feral Senses) is irrelevant
        if (level >= 19) {
            features.push(Ranger.level19(asis[4]))
        }
        if (level >= 20) {
            features.push(Ranger.level20())
        }
        return features
    }

    static level1(level: number, masteries: WeaponMastery[]): Feature {
        return new FeatureGroup([
            new AddClassLevel("Ranger", level),
            new WeaponMasteries(masteries),
            new SpellcastingFeat("Wis", Spellcaster.Half, level),
        ])
    }

    static level2(fightingStyle: Feature): Feature {
        return new FeatureGroup([fightingStyle])
    }

    static level3(subclass: RangerSubclass): Feature {
        if (subclass === "Hunter") {
            // Level 3 (Hunter's Lore) is irrelevant
            return new FeatureGroup([new ColossusSlayer()])
        } else {
            unreachable(subclass)
        }
    }

    static level4(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level5(): Feature {
        return new FeatureGroup([new ExtraAttack(2)])
    }

    static level8(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level12(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level16(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level17(): Feature {
        return new FeatureGroup([new PreciseHunter()])
    }

    static level19(asi: Feature): Feature {
        return new FeatureGroup([asi])
    }

    static level20(): Feature {
        return new FeatureGroup([new FoeSlayer()])
    }

    static hunterRanger(level: number): Character {
        const magicBonus = defaultMagicBonus(level)
        const character = new Character({
            stats: { Str: 10, Dex: 17, Con: 10, Int: 10, Wis: 16, Cha: 10 },
        })
        const features = Ranger.features({
            level,
            subclass: "Hunter",
            masteries: ["Topple"],
            fightingStyle: new Archery(),
            asis: [
                new AbilityScoreImprovement("Dex"),
                new AbilityScoreImprovement("Dex", "Wis"),
                new AbilityScoreImprovement("Wis"),
                new AbilityScoreImprovement("Wis"),
                new IrresistibleOffense("Dex"),
            ],
        })
        features.forEach((feat) => character.addFeature(feat))

        character.customTurn.addOperation(
            "action",
            new CastSpellOperation((character) => {
                if (
                    character.spells.highestSlot() < 4 ||
                    character.spells.isConcentrating()
                ) {
                    return undefined
                }
                return new SummonFey(character.spells.lowestSlot(4))
            })
        )
        character.customTurn.addOperation(
            "action",
            new CastSpellOperation((character) => {
                if (character.spells.isConcentrating()) {
                    return undefined
                }
                return new HuntersMark(character.spells.lowestSlot())
            })
        )
        character.customTurn.addOperation(
            "action",
            new DefaultAttackActionOperation(new Longbow({ magicBonus }))
        )

        return character
    }
}
